### ISSUE TITLE
Feature Request: Dissenter YouTube Comments

### ISSUE THREAD

**Issue by trymeouteh** [omarroth/invidious#569]
Please add a section and even an option the the account settings to be able to view the comments of a YouTube video from Dissenter. Dissenter is a new commenting section platform that allows you to comment on any URL including YouTube videos.

Or even a URL link that you can click on to view the dissenter comments for that YouTube Video URL would be a start.

**Comment by omarroth** [comment]
Is there any documentation on accessing comments for a given URL? From what I can tell they're available as HTML from `https://dissenter.com/discussion/begin?url={query}`, but there doesn't appear to be a way to get them as a format other than HTML, although there is gab-ai-inc/gab-dissenter-discussion#45.

May also make sense to open a request for Invidious support in gab-ai-inc/gab-dissenter-extension.

**Comment by trymeouteh** [comment]
I know they are working on a Wordpress plugin. I will make a feature request for them to allow redirecting of urls,

**Comment by omarroth** [comment]
Would you mind linking the issue here?

**Comment by moritzheiber** [comment]
I would just like to point out what kind of discussion you're inviting onto your platform when integrating with this "free speech" extension: https://dissenter.com/discussion/

If you needed further proof, here are articles relating to the "owner" of the extension that was posted as a reference:

https://arstechnica.com/tech-policy/2017/08/gab-the-right-wing-twitter-rival-just-got-its-app-banned-by-google/
https://www.wired.com/2016/09/gab-alt-rights-twitter-ultimate-filter-bubble/

Generally, I would love for Invidious as a project to not enable the spewing of hate, bigotry, racism and fascism under the mantle of "free speech" but take a firm stance against it, resulting in the refusal of integrating Dissenter and the closing of this issue.

**Comment by Gorcenski** [comment]
I recommend against such support, and I would, in addition to moral grounds, firmly caution against involving one's self with Gab in any way.

Gab has been identified as the platform where the Pittsburgh Tree of Life shooter was radicalized before his killings. [Source](https://www.nytimes.com/2018/10/28/us/gab-robert-bowers-pittsburgh-synagogue-shootings.html) In the hours after his mass shooting, one of his closest followers killed himself in a Washington, DC park. He was found with multiple magazines, and investigators believe he himself was considering a mass shooting. That man's brother was subsequently arrested by the FBI and charged with weapons violations. [Source](https://www.washingtonpost.com/local/public-safety/dc-man-arrested-on-gun-charge-after-relatives-alert-police-to-his-alleged-white-nationalist-outbursts/2018/11/13/e3a6d2fa-e771-11e8-b8dc-66cca409c180_story.html?utm_term=.c6c9232efe0b) Upon searching his premises, investigators found propaganda belonging to the terror group Atomwaffen Division. [Source](https://www.courtlistener.com/recap/gov.uscourts.dcd.201711/gov.uscourts.dcd.201711.5.7.pdf)

Gab has initially cooperated with federal authorities and has since been dealing with a subpoena by the Pennsylvania Attorney General's office. [Source](https://www.npr.org/2018/11/08/665648215/gab-server-subpoenaed-by-pennsylvania-attorney-general) The site is under constant surveillance for domestic terror threats.

Gab has been accused of a fraudulent SEC filing, allegedly inflating its user numbers in now-withdrawn financial filings by reporting accounts instead of active users. Comments from Gab's hosting provider have confirmed that their technical infrastructure is incapable of supporting the claimed number of users. [Source](https://www.splcenter.org/hatewatch/2019/02/14/no-way-gab-has-800000-users-web-host-says)

In addition to these matters, Gab's financial situation is dire and they have very little technical or financial support. There is no guarantee of ongoing development or support for an API, and from a purely technical standpoint one must consider the cost of maintaining a feature that can become dead weight at any time, like many of their past aspirational features.

Gab's active user base appears to be in sharp decline after controversy when some of their most popular users have been permanently banned for, inter alia, doxing and soliciting murder. [Source](https://www.newsweek.com/social-media-site-gab-bans-crying-nazi-killing-leftists-1367679)

In short, involvement with Gab or any of its associated products is an invitation for more trouble than it is worth, from a moral, legal, and technical basis, and I would firmly caution against it.

**Comment by omarroth** [comment]
From the above comments, I see 2 _**technical**_ issues with this feature:

> No API support

As mentioned above there currently isn't an API for integrating or customizing comments from Dissenter which makes implementation difficult, although there is an open issue gab-ai-inc/gab-dissenter-extension#45 which would help with this. Since Invidious has the same URL scheme as YouTube I would imagine adding support in the extension would be mostly straight-forward.

Additionally, for users to comment with Dissenter it would be necessary to authenticate with another service, which is a much larger feature request that Invidious currently doesn't support (see #364 and maybe #328).

> Low user count

For the amount of effort required I currently don't think it's feasible to add support for this directly. As mentioned above I think it would take less effort overall to implement in the Dissenter and would also provide a higher-quality experience.

Unless there is anything I failed to consider I would suggest closing in favor of gab-ai-inc/gab-dissenter-extension#86, although as mentioned in that issue I believe this is more of an edge-case, and would be better to request support for Invidious directly.

**Comment by trymeouteh** [comment]
Gab and Dissenter has been smeared by the media as a platform for "hate". The reason why I and many like Dissenter is because it is a free speech commenting platform. You can comment on any URL and that website cannot ban it. I can make a dissenter comment on youtube.com, invidio.us or a specific YouTube video and YouTube cannot remove the comment no matter how much youtube or invidious dislikes the comment.

Sure people can post things that have hatred in them but the reality is that the internet will always have hate, hate is an human emotion and is something every human being has and it is normal to experience hate and to have hatred. Not saying hate should be multiplied and encourage but it is a thing and to try to hide it, brush it under the carpet, censor it will not eliminate hate, the hate will still exist and will exist on the dark corners of the internet where it will not be challenged and then can breed and cause problems later on.

When you suppress speech, it will only increase violence since the healthy way for one to express oneself which is speech is not allowed, violence is the next option on expressing opposition. I do not want violence.

I am not suggestion dissenter comments to be enabled by default, make it optional just like reddit comments are optional. It is true that Dissenter has a low user count right now since it was released earlier this year.

I understand this would be big task to handle without an API, but this could be considered in the future if the users growth increases and if an API is created.

A simple suggestion would be to allow an URL to the dissenter comment page of a video which can look just like this...

[Dissenter](https://dissenter.com/discussion/begin?url=https://www.youtube.com/watch?v=QIqZNh78E-A)

**Comment by omarroth** [comment]
To be clear: I am not opposed to adding this as a feature. I would rather see questions or issues about implementation, not the quality of discussion or character of people involved. I've mistakenly done this before (see #1) and would rather not have the same happen here.

As mentioned above, this likely won't receive much attention on the Invidious side until there's proper support by Dissenter for integrating into other applications. My suggestion to request this directly in [gab-ai-inc/gab-dissenter-extension](https://github.com/gab-ai-inc/gab-dissenter-extension) is in the hope of reducing total workload for both projects.

**Comment by Gorcenski** [comment]
It's sad you've chosen fascism.

**Comment by trymeouteh** [comment]
> To be clear: I am not opposed to adding this as a feature. I would rather see questions or issues about implementation, not the quality of Dissenter discussion or character of people involved. I've mistakenly done this before with YouTube comments (see #1) and would rather not have the same happen here.
> 
> As mentioned above, this likely won't receive much attention on the Invidious side until there's proper support by Dissenter for integrating into other applications. My suggestion to request this directly in [gab-ai-inc/gab-dissenter-extension](https://github.com/gab-ai-inc/gab-dissenter-extension) is in the hope of reducing total workload for both projects.

I agree, to have a good Dissenter comment section added, Gab needs to make a good API that anyone can use. Personally I would prefer if Gab added a feature in Dissenter for users to block commenting on URLs or to redirect URLs so this way invidious would not have to do anything to get this feature.

**Comment by cloudrac3r** [comment]
Wanting to support and cultivate free speech is a very noble goal, but once you step back to consider that the term "free speech" is a shield for _literal fascists_ to hide behind, you have an ethical dilemma ahead of you.

As you know, Dissenter is created by Gab. Gab has all this innocent marketing about "free speech", which sounds very reasonable, but it does know who its true audience is, as evidenced by the front and centre image on https://dissenter.com/.

The image includes links to 4 videos, with their titles and thumbnails visible. The first one, named "CONFUSED: Angry Tommy Robinson HATER RANTS about LOVE". Tommy Robinson is an anti-Islam far-right political figure ([wikipedia](https://en.wikipedia.org/wiki/Tommy_Robinson_(activist))). If you actually watch the video, you find that it and its YouTube comment section are mocking the hater, therefore absolutely supporting hate against Islam.

"Journalism Is Overrun By Sociopathic Activist Bullies". I've seen a lot of far-right headlines in my time, and this is definitely one of them. Saying that media that you disagree with (i.e. all mainstream media!) is lies is a very common tactic used to make people doubt that what they are hearing is correct, and to make them search out "alternative" sources of dubious reliability, but ones that tell them what they want to hear.

Gab knows what it is. Dissenter knows what it is. It's a platform for hate speech against jews, muslims, foreigners, LGBT people and LGBT rights (which are, funnily enough, _human rights._)

"Well now, shouldn't this speech still be allowed?" you might ask. "After all, free speech means people have the rights to say this." It's true, they do have the right to say this. However, you also have the right to stop people from seeing this unless they specifically go looking for it (i.e. by installing the Dissenter browser extension, which is actually so controversial that it's banned from the Firefox Add-ons Store and the Google Extensions Store, meaning you have to actually install their own browser to get this extension.) You have the right to decide whether you want to promote this or not by either working to include a prominent button to display it on your website, or just not doing that and still allowing people to say what they want, without you explicitly promoting and supporting it.

So now the argument is this. Do you support hate speech, or do you support human rights?

**Comment by cloudrac3r** [comment]
In case my point wasn't clear enough, refusing to add support for Dissenter comments does not affect free speech, since it does not impact the ability of these people to speak whatsoever.

It also does not prevent people seeking out this speech if they want to, since the browser extension exists and is already in use by people who want to be part of that community.

**Comment by trymeouteh** [comment]
@cloudrac3r 

Your are right that by not having dissenter integrated into invidious will not affect free speech since you can get the extensions and when watching a video on invidious, you just have to click the Watch On YouTube link and then use the extensions to view the comments.

Now you made an argument saying "hate speech" should not be allowed. I disagree since what is defined as "hate speech" I can say something today that you will not find "hateful" but within a few years it can be considered "hate speech" by someone or some group. Hate speech is just a term that is used by someone or some group to silence voices, jokes, and options that they do not like. 

There are many opinions out there I find offensive and filled with rage, and yes they do exist on Dissenter and Gab. But nevertheless the "hate speech" is not harming anyone in any physical form. I do not want to live in a world where you cannot express your grievances, concerns or just express your thoughts, your opinions because some arbiter dictates what is hate and what is not hate.

Yes I am personally fine with "hate speech" since it is just speech, it is just words. I don't agree with all "hate speech" and I don't have to agree on every piece of "hate speech" to be for free speech and no one has to agree with everything to be for freedom of speech. Of course nazis and facists will use speech, who doesn't speak to get their options out there. Just because I think everyone, even nazis and facists should have a voice doesn't mean I am standing with them on their values and beliefs.

**Comment by justagoodperson** [comment]
Hilarious that there's literal antifa in this thread attacking Gab and telling the devs to stay away from them. @Gorcenski you may want to let your obvious conflict of interest be apparent as opposed to pretending to be making an impartial good faith assessment of the developer's pros and cons of incorporating Dissenter into their application.

My personal opinion is that unless you are willing to hack in an iframe that probably waiting until there is some kind of REST API support for Dissenter comments is prudent. My VirtualBox installation seems to be bugged so I can't boot up docker and check this project out atm to confirm that a hacked in iframe wouldn't look or feel that great from a UI perspective but that's just what my intuition tells me. 

Likewise, considering that the browser extension is OS I agree with @omarroth that this discussion should be on a technical and UI perspective as opposed to a crap-flinging contest over the character of Gab as a company. Technically it does make at least some sense that an open-source comments extension would be a good fit for this project, seeing that in the future someone could create their own comments-based community and plug it into this project as a workaround for YouTube's built-in comment system. But yeah I agree that for the time being probably OAuth + Rest API would be the best implementation of which the REST API is not supported atm to the best of my knowledge.

**Comment by Perflyst** [comment]
Can we please stop to talk about the ethical things here? I think @omarroth can decide what he wants to implement or not, he is the maintainer of this project.

As said here https://github.com/omarroth/invidious/issues/569#issuecomment-499104121 it seems to be not easily possible from the technical side.

**Comment by elypter** [comment]
>Hilarious that there's literal antifa

up until now i had the assumption that people here really just want free speech but this is a clear right extremist delusional wording. antifa literally means anti fascism and the conspiracy some imagine is delusional right wing evangelist propaganda. Usually i am all for allowing all sorts of compatibilities even if i boldly disagree with a projects direction or users. however i also understand that people do not want a platform that is mostly bad content. i am still not opposed to dissenter and i like the idea of an independend commenting system. i had this idea in the past as well but lets not focus on it too much. there are other comment systems that are worth a look and might be easier to integrate

**Comment by Gorcenski** [comment]
> Hilarious that there's literal antifa in this thread

I do not belong to any antifa organization and have testified, under oath, to such.

> your obvious conflict of interest

I have no stake, financial or otherwise, in Gab or any of its competitors.

> Can we please stop to talk about the ethical things here

All technical decisions, including the decision to avoid speaking about them out loud, are also ethical matters.

> crap-flinging contest over the character of Gab

I sit on the advisory board of two university laboratories studying domestic extremist violence. The “crap” I slung included half a dozen links, including primary source material. The maintainer can choose to ignore it or accept it, but it is well-researched.

> however i also understand that people do not want a platform that is mostly bad content. i am still not opposed to dissenter

You all can barely cope with a small handful of polite, softly-spoken, well researched comments on a github issue designed for this without resorting to libel, and you want to extend that functionality to everything the project touches? Why should you feel ok extending this to non-consenting creators on youtube when you can barely handle it in a consent forum designed tor it yourselves?

**Comment by Gorcenski** [comment]
It is also not without notice that in a discussion supposedly about free speech, some are trying to restrict discussions of certain topics, and resorting to libel to do so.

One might conclude that the supposed free speech argument isn’t really genuine.

**Comment by MrPetovan** [comment]
If @omarroth doesn't even take a symbolic stand against Gab and Dissenter, I cannot in good faith recommend Invidious anymore like I stopped recommending HookTube before for the same reason.

**Comment by Perflyst** [comment]
Than do so. Let the people use YouTube who tracks its users and doesnt even care about gab. 

This makes no sense dude

**Comment by MrPetovan** [comment]
All things considered, they're marginally better off tracked on Youtube than exposed to unmoderated Dissenter comments. I don't care if people are using Dissenter, I care about non-users being potentially exposed to it on Invidious.

People already using Dissenter don't need this integration. People not using it don't need to know about it. This is what at stake here.

**Comment by justagoodperson** [comment]
Out of respect for @omarroth's request and his time I will hold back on continuing to use this thread to discuss nontechnical considerations, unless he gives me the greenlight and I will give some more rejoinders in what is now this debate, so up to you @omarroth on whether you'd like to cap this discussion or have it play out more here.

**Comment by trymeouteh** [comment]
Like I said above I did not want this becoming a debate but that cat is out of the bag now. Most of this open Issue is a debate, not development progress.

@omarroth seems to have taken a neutral stance on this and I am actually am thankful for it. I would rather him be neutral than praising or condemning the Dissenter platform.

@omarroth said it will not happen until there is a good API made to integrate into Invidious which is fair enough and he does have a good point on low user count since it is a new platform but I do think in time Dissenter will grow to be a major commenting platform. I don't think @omarroth is reading this and if he is he is rolling his eyes and or laughing at some of our comments. He will not develop this feature for time and if no API for Dissenter is made it may never get made, and if that is the case I am fine with it. I cannot have every feature request of mine come to life.

**Comment by trymeouteh** [comment]
> I sit on the advisory board of two university laboratories studying domestic extremist violence. The “crap” I slung included half a dozen links, including primary source material. The maintainer can choose to ignore it or accept it, but it is well-researched.

No offense, but I know that universities have a strong bias and the research, and "facts" are spun to fit an narrative. I do not trust any university with any studies that relate to some political topic or issue especially "monitoring hate". And I am not the only one. I know that the research many universities put out is either missing chunks of data, or presented in a way to appear there is a problem online or in society of "extreme violence". I trust my eyes and ears and I know that many of these studies do not reflect on reality.

I have not went to university and I will never go to university simply because I will probably be expelled for having ideas, expressing my opinions and by being expelled, losing thousands of dollars and not getting the degree I went in for. It is a shame I think since I am not a violent person, I am a libertarian. And also university degrees are losing value every ear since many university student end up in low wage jobs after spending thousands of dollars on high education which is a sad reality.

I want peace, love and happiness but I do understand that when there is peace, love and happiness in this world there is also war, hate and sadness and that is just the world we live in, These emotions, opinions and feelings exist and it is natural to feel this way in life, it is part of being human. I have been mad before at others, There are people out there I do not like. And to police this behavior on society is toxic and to try to control human emotions will only bottle up the anger, rage and it will one day erupt and explode and when it does, it will be violent and brutal since society will have removed all healthy ways to deal with negative emotions such as freedom of speech being a healthy and safe way of expressing such disagreements.

Your studies may say that there is extremists online, sure there is. There will always be that. The reason many of these extremists people are online expressing themselves is because they have been shunned from the physical public square and many silenced on their university campus. If you remove them from the internet and they do not have anywhere else to go to express themselves, do not be surprised if these extremist take violent action towards those they dislike in the physical world. And that is something I do not want to see happen.

**Comment by JPLeBreton** [comment]
> I have not went to university and I will never go to university simply because I will probably be expelled for having ideas, expressing my opinions and by being expelled, losing thousands of dollars and not getting the degree I went in for.

Can you give some specific examples of ideas and opinions you hold that you think you would be expelled for?

**Comment by omarroth** [comment]
Since this is drifting off-topic I'm locking discussion for now. As already mentioned I believe this is better handled in [gab-ai-inc/gab-dissenter-extension](https://github.com/gab-ai-inc/gab-dissenter-extension) until there's better support for third-party applications.
